jerry-core/ecma/base/ecma-helpers-string.c

```diff
@@ -2482,6 +2507,185 @@ ecma_string_trim (const ecma_string_t *string_p) /**< pointer to an ecma string
+                               const lit_utf8_size_t data_size) /**< size of the data */
+{
+  lit_utf8_byte_t *dest_p = ecma_stringbuilder_grow (builder_p, data_size);
+  memcpy (dest_p, data_p, data_size);
+} /* ecma_stringbuilder_append_raw */
+
+/**
+ * Append an ecma_char_t to a string builder
+ */
+void
+ecma_stringbuilder_append_char (ecma_stringbuilder_t *builder_p, /**< string builder */
+                                const ecma_char_t c) /**< ecma char */
+{
+  const lit_utf8_size_t size = (lit_utf8_size_t) lit_char_get_utf8_length (c);
+  lit_utf8_byte_t *dest_p = ecma_stringbuilder_grow (builder_p, size);
+
+  lit_char_to_utf8_bytes (dest_p, c);
+} /* ecma_stringbuilder_append_char */
+
+/**
+ * Finalize a string builder
+ *
+ * @return the created string from the underlying buffer
+ */
+ecma_string_t *
+ecma_stringbuilder_finalize (ecma_stringbuilder_t *builder_p) /**< string builder */
+{
+  ecma_string_t *string_p;
+  ecma_stringbuilder_header_t *header_p = builder_p->header_p;
+  const lit_utf8_size_t string_size = ECMA_STRINGBUILDER_STRING_SIZE (header_p);
+  lit_utf8_byte_t *string_begin_p = ECMA_STRINGBUILDER_STRING_PTR (header_p);
+
+  if (JERRY_UNLIKELY (ecma_is_string_special (string_begin_p,
+                                              string_size,
+                                              &string_p)))
+  {
+    ecma_stringbuilder_destroy (builder_p);
+    return string_p;
+  }
+
+  string_p = (ecma_string_t *) header_p;
+  const lit_string_hash_t hash = lit_utf8_string_calc_hash (string_begin_p, string_size);
+  const lit_utf8_size_t length = lit_utf8_string_length (string_begin_p, string_size);
+
+  if (JERRY_LIKELY (string_size < UINT16_MAX))
+  {
+#if ENABLED (JERRY_MEM_STATS)
+    jmem_stats_allocate_string_bytes (header_p->current_size);
+#endif /* ENABLED (JERRY_MEM_STATS) */
+
+    string_p->refs_and_container = ECMA_STRING_CONTAINER_HEAP_UTF8_STRING | ECMA_STRING_REF_ONE;
+    string_p->u.utf8_string.size = (uint16_t) string_size;
+    string_p->u.utf8_string.length = (uint16_t) length;
+    string_p->hash = hash;
+
+    return string_p;
+  }
+
+  const size_t required_size = header_p->current_size - sizeof (ecma_string_t) + sizeof (ecma_long_string_t);
+  header_p = jmem_heap_realloc_block (header_p, header_p->current_size, required_size);
+  memmove (((ecma_long_string_t *) header_p) + 1, string_begin_p, string_size);
+
+  string_p->refs_and_container = ECMA_STRING_CONTAINER_HEAP_LONG_UTF8_STRING | ECMA_STRING_REF_ONE;
+  string_p->u.long_utf8_string_size = string_size;
+  string_p->hash = hash;
+
+  ecma_long_string_t *long_string_p = (ecma_long_string_t *) string_p;
+  long_string_p->long_utf8_string_length = length;
+
+#if ENABLED (JERRY_MEM_STATS)
+  jmem_stats_allocate_string_bytes (required_size);
+#endif /* ENABLED (JERRY_MEM_STATS) */
+
+  return string_p;
+} /* ecma_stringbuilder_finalize */
+
+/**
+ * Destroy a string builder
```

Comment:
When should this be called? Should it be called after calling finalize?

Reply:
This can be used to free a builder that is no longer needed, for example when we encounter an error. I will update the comment to be a bit more descriptive.